### ISSUE TITLE
SSH Server no longer running?

### ISSUE THREAD

**Issue by hannorein** [jankais3r/Unistellar-eVscope-research#3]
Unless I'm missing something, it seems that the SSH server is no longer running in the latest firmware version. 😢

**Comment by jankais3r** [comment]
Thank you for the note, I will try to find time during the weekend to look into it. In the meantime, could you search the firmware image for `dropbear`, just to see if the server daemon is still present? If it is, activating it will be a lot less work than adding it in.

**Comment by hannorein** [comment]
Yes, I still see a reference for dropbear. But I'm not sure how to enable it. The only open ports are the ones you mentioned plus a new one, 13010 (but that's not the SSH server either).

**Comment by hannorein** [comment]
I'd be happy to look into this, but I'd need some suggestions on how to approach this. I've got the firmware image and can change the SSH key as you described. But is it possible to unpack the firmware image to explore the file system and look for the right configuration file?

**Comment by jankais3r** [comment]
Hi, just wanted to drop you a quick update message. I spent a half day looking into this, but I do not have a solution at this point. You are right that we need to find a way to "unpack" the firmware image. My only workable solution right now is to downgrade the firmware to a version which still had SSH enabled. I have 4 different historical versions and only the last one has SSH disabled.

**Comment by hannorein** [comment]
Oh wow. Thanks for looking into it! 

That's a shame. At least it looks like the made it possible to access the raw data by downloading it from their servers. Still, not an ideal workflow...

**Comment by jankais3r** [comment]
So I again had a bit of time on my hands and I figured out how to extract filesystem from the firmware image… I have not, so far, succeeded in making the raspberry successfully boot after I made edits to the image, but I am fairly confident that there is a way to do that.

A quick recap for future reference: To get the FW file you either have to take off the cover of the telescope, take out the SD card and copy the `evscope.fw` file from the `datasys` partition, or you can extract it from the mobile app. For example on macOS you can easily download the IPA using `ipatool download -b com.unistellar.evscope.iOS.store`. Once you have the IPA file, unzip it to get to the FW image.

There are two incompatibilities with a regular `extract-vmlinux` script that we have to solve.

**1) ARM compatiblity**
By default, the `extract-vmlinux` script doesn't work with ARM kernel images. There is a [patch](https://patchwork.kernel.org/project/linux-kbuild/patch/1453810236-2082-1-git-send-email-rogershimizu@gmail.com/) available, but it hasn't been merged into mainline Linux, so we have to manually apply it.
```
curl https://patchwork.kernel.org/project/linux-kbuild/patch/1453810236-2082-1-git-send-email-rogershimizu@gmail.com/raw/ -o patch.patch
curl https://raw.githubusercontent.com/torvalds/linux/master/scripts/extract-vmlinux -o extract-vmlinux
patch extract-vmlinux patch.patch
```

If you got `Hunk #3 succeeded at 58 (offset -3 lines).` as a result of the above commands, you now have an ARM-compatible kernel extractor.

But if you run it now, it will not work yet.
```
% ./extract-vmlinux evscope.fw
tr: Illegal byte sequence
tr: Illegal byte sequence
tr: Illegal byte sequence
tr: Illegal byte sequence
tr: Illegal byte sequence
tr: Illegal byte sequence
tr: Illegal byte sequence
extract-vmlinux: Cannot find vmlinux.
```

**2) Locale**

I can't find the exact stackoverflow post that ultimately lead me to this solution, but the problem is caused by macOS' default terminal locale. We need to manually set it to POSIX's default LANG=C, then it will work:

`LANG=C ./extract-vmlinux evscope.fw > evscope_extracted.bin`

Great, now we have an extracted kernel image. How to get a filestem out of it? With `binwalk`.
Run `binwalk -e evscope_extracted.bin` to extract the files, or omit the `-e` if you only want to get a list of files without extracting them.

You will receive a file called `93645C.cpio`, which holds the whole filesystem. MacOS can extract cpio archives just by double clicking, or you can extract it from terminal by running `cat 93645C.cpio | cpio -idv`.

This week Unistellar pushed a 1.6 update to the eVscope's filesystem and I ran a quick comparison between 1.5 and 1.6 versions to see what has changed. See the attached CSV for comparison.
[FS_diff.csv](https://github.com/jankais3r/Unistellar-eVscope-research/files/8634535/FS_diff.csv)
Some files were just bumped to new version (same path, different md5 hash), while some files are new and some were removed.

You were right that `dropbear` (or any other SSL server) is no longer included, however there is still an `authorized_keys` file that we could utilize if we added `dropbear` to the image. However, I am currently thinking to maybe just add a reverse shell command to one of the bash scripts that is being auto-launched on boot… I'll have to dig around, but I will post any findings here.

**Comment by Freyr86** [comment]
Good morning,
Did you work on it again? I wanted to see if it was possible to retrieve the images directly in SSH but I haven't yet found a way to activate it. As you said it is no longer in the new firmware version.

**Comment by ahezard** [comment]
Hi I tried to open my eVscope 1 for the first time. I was happyto be able to upgrade the sd card from 16GB to 64GB (with partition extension). This is already cool since the 16 GB was limiting my observation time. I am trying to understand how to access the scope. Did you look into serial access instead of ssh for newest firmware?

The cmdline.txt (fw 4.3.5) contains 
rootwait console=tty1 console=ttyS0,115200 console=ttyUSB0,115200 logo.nologo
Do you think one of the usb port allows serial over usb? ttyUSB0 is interesting
